macros/src/main/scala-3/enumeratum/EnumMacros.scala

```diff
@@ -0,0 +1,212 @@
+         """.stripMargin
+      )
+    } else {
+      Ref(repr.typeSymbol.companionModule).asExprOf[M]
+    }
+  }
+
+  /** Makes sure that we can work with the given type as an enum:
+    *
+    * Aborts if the type is not sealed.
+    *
+    * @tparam T
+    *   the `Enum` type
+    */
+  private[enumeratum] def validateType[T](using q: Quotes, tpe: Type[T]): q.reflect.TypeRepr = {
+    import q.reflect.*
+
+    val repr = TypeRepr.of[T](using tpe)
+
+    if (!repr.classSymbol.exists(_.flags is Flags.Sealed)) {
+      report.errorAndAbort(
+        "You can only use findValues on sealed traits or classes"
+      )
+    }
+
+    repr
+  }
+
+  /** Returns a sequence of symbols for objects that implement the given type
+    *
+    * @tparam T
+    *   the `Enum` type
+    * @param tpr
+    *   the representation of type `T` (also specified by `tpe`)
+    */
+  private[enumeratum] def enclosedSubClasses[T](q: Quotes)(
+      tpr: q.reflect.TypeRepr
+  )(using tpe: Type[T]): List[q.reflect.TypeRepr] = {
+    import q.reflect.*
+
+    // TODO: Use SumOf?
+    given quotes: q.type = q
+
+    @annotation.tailrec
+    def isObject(sym: Symbol, ok: Boolean = false): Boolean = {
+      if (!sym.flags.is(Flags.Module)) {
+        val owner = sym.maybeOwner
+
+        if (
+          owner == defn.RootClass ||
+          owner.flags.is(Flags.Package) ||
+          owner.isAnonymousFunction
+        ) {
+          report.errorAndAbort(
+            // Root must be a module to have same behaviour
+            "The enum (i.e. the class containing the case objects and the call to `findValues`) must be an object"
+          )
+        }
+
+        false
+      } else {
+        val owner = sym.maybeOwner
+
+        if (!owner.exists || owner == defn.RootClass || owner.isTerm) {
+          if (
+            sym.flags.is(Flags.Module) && !sym.flags.is(Flags.Package) &&
+            !sym.fullName.startsWith(tpr.typeSymbol.companionModule.fullName)
+          ) {
+            // See EnumSpec#'should return -1 for elements that do not exist'
+
+            report.warning(
+              s"The entry '${sym.fullName}' must be defined in the enum companion"
+            )
+
+            false
+          } else {
+            ok
+          }
+        } else {
+          isObject(sym.maybeOwner, true)
+        }
+      }
+    }
+
+    type IsEntry[E <: T] = E
+
+    @annotation.tailrec
+    def subclasses(
+        children: List[Tree],
+        out: List[TypeRepr]
+    ): List[TypeRepr] = {
+      val childTpr = children.headOption.collect {
+        case tpd: Typed =>
+          tpd.tpt.tpe
+
+        case vd: ValDef =>
+          vd.tpt.tpe
+
+        case cd: ClassDef =>
+          cd.constructor.returnTpt.tpe
+
+      }
+
+      childTpr match {
+        case Some(child) => {
+          child.asType match {
+            case '[IsEntry[t]] => {
+              val tpeSym = child.typeSymbol
+              // TODO: Check is subtype (same in Scala2?)
```

Comment:
@cchantep :c

Curious if this is already verified by the type match + compiler because of the type alias above of `IsEntry[E <: T]`?

Reply:
Updated